### ISSUE TITLE
A deep dive into Linux namespaces

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4745]
A deep dive into Linux namespaces<br>
<br>
https://ift.tt/2xiGeGZ<br>
<br>
<br>
<br>
<div>
<img title="A deep dive into Linux namespaces – Chord Simple" src="http://ifeanyi.co/images/default-thumb.png"><br><div>
<div> <section><header><h3>Overview</h3> </header></section><p>In this series of posts we will look closely at one of the main ingredients in a <a href="https://www.docker.com/resources/what-container" target="_blank">container</a> - Namespaces. In the process, we will create a simpler clone of the <code>docker run</code> command - our very own program that will take as input a command (along with it’s arguments if any) and spin up a <em>container</em> process to run it, isolated from the rest of the system similar to how you would <code>docker run</code> it <a href="https://docs.docker.com/engine/reference/run/" target="_blank">from an image</a>.</p> <h2>What is a namespace?</h2> <p>A <a href="http://man7.org/linux/man-pages/man7/namespaces.7.html" target="_blank">Linux namespace</a> is an abstraction over resources in the operating system. We can think of a namespace as a box. Inside this box are these system resources, which ones exactly depend on the box’s (namespace’s) type. There are currently 7 types of namespaces <code>Cgroup</code>, <code>IPC</code>, <code>Network</code>, <code>Mount</code>, <code>PID</code>, <code>User</code>, <code>UTS</code>.</p> <p>For instance, the <code>Network</code> namespace encapsulates system resources related to networking such as network interfaces (e.g <code>wlan0</code>, <code>eth0</code>), route tables etc, the <code>Mount</code> namespace encapsulates files and directories in the system, <code>PID</code> contains process IDs and so on. So two instances of a <code>Network</code> namespace <code>A</code> and <code>B</code> (corresponding to two boxes of the same type in our analogy) can contain different resources - maybe <code>A</code> contains <code>wlan0</code> while <code>B</code> contains <code>eth0</code> and a different route table copy.</p> <p>Namespaces aren’t some addon feature or library that you need to apt install, they are provided by the Linux kernel itself and already are a prerequisite to run any process on the system. At any given moment, any process <code>P</code> belongs to exactly one instance of each namespace type - so when it needs to say, update the route table on the system, Linux shows it the copy of the route table of the namespace to which it belongs at that moment.</p> <h2>What is it good for?</h2> <p>Absolutely nothi… just kidding. One good thing with boxes is that you can add and remove stuff from one box and it will not affect the content of other boxes. That’s the same idea here with namespaces - a process <code>P</code> can go crazy and <code>sudo rm -rf /</code> but another process <code>Q</code> that belongs to a different <code>Mount</code> namespace will be unaffected since they’re using distinct copies of those files.</p> <p>Note though that a resource encapsulated within a namespace doesn’t necessarily mean that it’s a unique copy. In a number of cases, either by design or as a security hole, two or more namespaces will contain the same copy, e.g of the same file, so that changes made to that file in one <code>Mount</code> namespace will in fact be visible in all other <code>Mount</code> namespaces that also reference it. For this reason, we will retire our box analogy here since an item cannot simultaneously exist in two distinct boxes 😞.</p> <h2>Unsharing is caring</h2> <p>We can see the namespaces that a process belongs to! In typical Linux fashion, they’re exposed as files under the directory <code>/proc/$pid/ns</code> for a given process with process id <code>$pid</code>:</p> <figure><pre><code><span>$ </span><span>ls</span> <span>-l</span> /proc/<span>$$</span>/ns total 0 lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 cgroup -&gt; cgroup:[4026531835] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 ipc -&gt; ipc:[4026531839] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 mnt -&gt; mnt:[4026531840] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 net -&gt; net:[4026531957] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 pid -&gt; pid:[4026531836] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 user -&gt; user:[4026531837] lrwxrwxrwx 1 iffy iffy 0 May 18 12:53 uts -&gt; uts:[4026531838]</code></pre></figure><p>You can open a second terminal and run the same command and it should give you the exact same output - this is because, as we mentioned earlier, a process must belong to some namespace and unless we explicitly specify which ones, Linux adds it as a member to the default namespaces. <br><br> Let’s meddle in this a bit. In the second terminal we can run something like:</p> <figure><pre><code><span>$ </span>hostname iffy <span>$ </span><span>sudo </span>unshare <span>-u</span> bash <span>$ </span><span>ls</span> <span>-l</span> /proc/<span>$$</span>/ns lrwxrwxrwx 1 root root 0 May 18 13:04 cgroup -&gt; cgroup:[4026531835] lrwxrwxrwx 1 root root 0 May 18 13:04 ipc -&gt; ipc:[4026531839] lrwxrwxrwx 1 root root 0 May 18 13:04 mnt -&gt; mnt:[4026531840] lrwxrwxrwx 1 root root 0 May 18 13:04 net -&gt; net:[4026531957] lrwxrwxrwx 1 root root 0 May 18 13:04 pid -&gt; pid:[4026531836] lrwxrwxrwx 1 root root 0 May 18 13:04 user -&gt; user:[4026531837] lrwxrwxrwx 1 root root 0 May 18 13:04 uts -&gt; uts:[4026532474] <span>$ </span>hostname iffy <span>$ </span>hostname coke <span>$ </span>hostname coke</code></pre></figure><p>The <code>unshare</code> command runs a program (optionally) in a new namespace. The <code>-u</code> flag tells it to run <code>bash</code> in a new <code>UTS</code> namespace. Notice how our new <code>bash</code> process points to a different <code>uts</code> file while all others remain the same.</p> <blockquote> <p>Creating new namespaces usually requires superuser access. From now on, we will assume that both <code>unshare</code> or our implementation are run with <code>sudo</code>.</p> </blockquote> <p>One implication of what we just did is that we can now change the system’s <code>hostname</code> from within our new <code>bash</code> process and it won’t affect any other process in the system. You can verify this by running <code>hostname</code> in the first shell or a new one and seeing that the hostname hasn’t changed there.</p> <h2>But like, what is a container though?</h2> <p>Hopefully, now you have some idea of what a namespace can do. You might guess that containers are fundamentally ordinary processes with different namespaces from other processes and you’d be correct. In fact a quote, unquote container doesn’t have to belong to a unique namespace for each type - it can share some of them.</p> <p>For instance, when you <code>docker run --net=host redis</code>, all you do is tell docker to not create a new <code>Network</code> namespace for the <code>redis</code> process, and as we saw, Linux will add that process as a member of the default <code>Network</code> namespace just like every other regular process. So the redis process is exactly like everyone else from a networking perspective. Networking isn’t special here,<code>docker run</code> let’s you do this customization for most namespaces. This begs the question of what even is a container? Is a process that shares all but one namespace still a container? ¯\_(ツ)_/¯ <br> Usually containers come with the notion of <strong>isolation</strong>, achieved through namespaces - the smaller the number of namespaces and resources that a process shares, the more isolated the process is and that’s all that really matters.</p> <h2>Isolate</h2> <p>In the remainder of this post, we will lay the ground work for our program that we will call <code>isolate</code>. <code>isolate</code> takes a command as arguments and runs that command in a new process isolated from the rest of the system and within its very own namespaces. In the coming posts, we will look at adding support for individual namespaces when <code>isolate</code> spins up the command process.</p> <p>In terms of scope, we will focus on the <code>User</code>, <code>Mount</code>, <code>PID</code> and <code>Network</code> namespaces. The rest are relatively trivial to implement once we’re done (in fact, we add <code>UTS</code> support in the initial implementation here) and <code>Cgroup</code> for example is only interesting from a perspective that is out of scope of this series (studying <code>cgroups</code> - the <a href="http://man7.org/linux/man-pages/man7/cgroups.7.html" target="_blank">other ingredient in containers</a> that is used to control how much of a resource a process is allowed to use).</p> <p>Namespaces can get complex real quick so there are lots of different paths we can take while studying each namespace but we can’t take them all. We will only discuss the paths that are relevant to the program that we’re building. Each post will start off with some experiments on the namespace in question within a terminal in an attempt to understand the interactions involved in setting up that namespace. After this we will already have an idea of what we want to accomplish and will then follow up with a corresponding implementation in <code>isolate</code>.</p> <blockquote> <p>To avoid bombarding the posts with code, we will not include things like helper functions that are not necessary to understand the implementation. You can find the full source code <a href="https://github.com/iffyio/isolate" target="_blank">here on Github</a>.</p> </blockquote> <h2>Implementation</h2> <blockquote> <p>The source code for this post <a href="https://github.com/iffyio/isolate/tree/part-1" target="_blank">can be found here</a>.</p> </blockquote> <p>Our <code>isolate</code> implementation will initially be a simple program that reads a command path from stdin and clones a new process that executes the command with the specified arguments. The cloned command process will run in its own <code>UTS</code> namespace just like we did with <code>unshare</code> earlier. In later posts, we will see that namespaces do not necessarily work (or even provide isolation) out of the box and we will need to do some setup after creating them (but before executing the actual command) in order for the command to truly run in isolation.</p> <p>This namespace creation-setup combo will require some co-operation between the main <code>isolate</code> process and the child command process. As a result, part of the ground work here will be to setup a communication channel between both processes - we will use a <a href="https://www.tldp.org/LDP/lpg/node11.html" target="_blank">Linux pipe</a> due to its simplicity given our use case.</p> <p><br> We have three things to do:</p> <ol>
<li>Create the main <code>isolate</code> process that reads from stdin.</li> <li>Clone a new process that will run the command in a new <code>UTS</code> namespace.</li> <li>Set up a pipe so that the command process begins the command execution only after it receives a signal from the main process that the namespace setup is done.</li> </ol>
<p>Here is the main process:</p> <figure><pre><code><span>int</span> <span>main</span><span>(</span><span>int</span> <span>argc</span><span>,</span> <span>char</span> <span>**</span><span>argv</span><span>)</span> <span>{</span> <span>struct</span> <span>params</span> <span>params</span><span>;</span> <span>memset</span><span>(</span><span>&amp;</span><span>params</span><span>,</span> <span>0</span><span>,</span> <span>sizeof</span><span>(</span><span>struct</span> <span>params</span><span>));</span> <span>parse_args</span><span>(</span><span>argc</span><span>,</span> <span>argv</span><span>,</span> <span>&amp;</span><span>params</span><span>);</span> <span>// Create pipe to communicate between main and command process. </span> <span>if</span> <span>(</span><span>pipe</span><span>(</span><span>params</span><span>.</span><span>fd</span><span>)</span> <span>&lt;</span> <span>0</span><span>)</span> <span>die</span><span>(</span><span>"Failed to create pipe: %m"</span><span>);</span> <span>// Clone command process. </span> <span>int</span> <span>clone_flags</span> <span>=</span> <span>SIGCHLD</span> <span>|</span> <span>CLONE_NEWUTS</span> <span>;</span> <span>int</span> <span>cmd_pid</span> <span>=</span> <span>clone</span><span>(</span><span>cmd_exec</span><span>,</span> <span>cmd_stack</span> <span>+</span> <span>STACKSIZE</span><span>,</span> <span>clone_flags</span><span>,</span> <span>&amp;</span><span>params</span><span>);</span> <span>if</span> <span>(</span><span>cmd_pid</span> <span>&lt;</span> <span>0</span><span>)</span> <span>die</span><span>(</span><span>"Failed to clone: %m</span><span>\n</span><span>"</span><span>);</span> <span>// Get the writable end of the pipe. </span> <span>int</span> <span>pipe</span> <span>=</span> <span>params</span><span>.</span><span>fd</span><span>[</span><span>1</span><span>];</span> <span>// Some namespace setup will take place here ... </span> <span>// Signal to the command process we're done with setup. </span> <span>if</span> <span>(</span><span>write</span><span>(</span><span>pipe</span><span>,</span> <span>"OK"</span><span>,</span> <span>2</span><span>)</span> <span>!=</span> <span>2</span><span>)</span> <span>die</span><span>(</span><span>"Failed to write to pipe: %m"</span><span>);</span> <span>if</span> <span>(</span><span>close</span><span>(</span><span>pipe</span><span>))</span> <span>die</span><span>(</span><span>"Failed to close pipe: %m"</span><span>);</span> <span>if</span> <span>(</span><span>waitpid</span><span>(</span><span>cmd_pid</span><span>,</span> <span>NULL</span><span>,</span> <span>0</span><span>)</span> <span>==</span> <span>-</span><span>1</span><span>)</span> <span>die</span><span>(</span><span>"Failed to wait pid %d: %m</span><span>\n</span><span>"</span><span>,</span> <span>cmd_pid</span><span>);</span> <span>return</span> <span>0</span><span>;</span> <span>}</span></code></pre></figure><p>Check out <code>clone_flags</code> that we pass to our <code>clone</code> invocation, See how dead simple it is to create a new process in its own namespace? All we have to do is set the flag for the namespace type (the <code>CLONE_NEWUTS</code> flag corresponds to the <code>UTS</code> namespace) and Linux takes care of the rest. <br><br> Next, the command process waits for a signal before exec-ing the command:</p> <figure><pre><code><span>static</span> <span>int</span> <span>cmd_exec</span><span>(</span><span>void</span> <span>*</span><span>arg</span><span>)</span> <span>{</span> <span>// Kill the cmd process if the isolate process dies. </span> <span>if</span> <span>(</span><span>prctl</span><span>(</span><span>PR_SET_PDEATHSIG</span><span>,</span> <span>SIGKILL</span><span>))</span> <span>die</span><span>(</span><span>"cannot PR_SET_PDEATHSIG for child process: %m</span><span>\n</span><span>"</span><span>);</span> <span>struct</span> <span>params</span> <span>*</span><span>params</span> <span>=</span> <span>(</span><span>struct</span> <span>params</span><span>*</span><span>)</span> <span>arg</span><span>;</span> <span>// Wait for 'setup done' signal from the main process. </span> <span>await_setup</span><span>(</span><span>params</span><span>-&gt;</span><span>fd</span><span>[</span><span>0</span><span>]);</span> <span>char</span> <span>**</span><span>argv</span> <span>=</span> <span>params</span><span>-&gt;</span><span>argv</span><span>;</span> <span>char</span> <span>*</span><span>cmd</span> <span>=</span> <span>argv</span><span>[</span><span>0</span><span>];</span> <span>printf</span><span>(</span><span>"===========%s============</span><span>\n</span><span>"</span><span>,</span> <span>cmd</span><span>);</span> <span>if</span> <span>(</span><span>execvp</span><span>(</span><span>cmd</span><span>,</span> <span>argv</span><span>)</span> <span>==</span> <span>-</span><span>1</span><span>)</span> <span>die</span><span>(</span><span>"Failed to exec %s: %m</span><span>\n</span><span>"</span><span>,</span> <span>cmd</span><span>);</span> <span>die</span><span>(</span><span>"¯</span><span>\\</span><span>_(ツ)_/¯"</span><span>);</span> <span>return</span> <span>1</span><span>;</span> <span>}</span></code></pre></figure><p>Finally we can try it out:</p> <figure><pre><code><span>$ </span>./isolate sh <span>===========</span><span>sh</span><span>============</span> <span>$ </span><span>ls </span>isolate isolate.c isolate.o Makefile <span>$ </span>hostname iffy <span>$ </span>hostname coke <span>$ </span>hostname coke <span># Verify in a new terminal that hostname hasn't been updated</span></code></pre></figure><p>Currently, <code>isolate</code> is a little bit more than a program that just forks off a command (we do have the <code>UTS</code> thing going for us). In the <a href="http://ifeanyi.co/posts/linux-namespaces-part-2/" target="_blank">next post</a>, we take it a step further by looking at <code>User</code> namespaces and have <code>isolate</code> run the command in its own <code>User</code> namespace. There, we will see that we actually need to do some work in order to have a usable namespace in which the command can run.</p> <hr>
<footer><p><strong>A deep dive into Linux namespaces</strong> was published on <time datetime="2019-06-27T00:00:00+00:00">June 27, 2019</time>.</p> </footer>
</div> </div>
</div><br>
<br>
<br>
<br>
<br>
<br>
via Chord Simple <br>
<br>
July 11, 2019 at 03:59PM
